# <span style='color:darkgreen'> Python for Data Analysis</span>
In this notebook, we will look into how to analyze data in Python using different libraries, including `pandas`, `numpy`, `matplotlib`, `seaborn`

We will look into different steps in analysing the data,
- **STEP 1**: starting from **data acquisition or import**,
- **STEP 2**: then doing **pre-processing** of the data,
- **STEP 3**: then doing **exploratory data analysis** to understand more about the data and different ways to **visualize** the outcome.
- **STEP 4**: Lastly, let's build a simple **ML model** for **predicting** the car price.

<hr>

## <span style='color:darkred'> **STEP 1: Data Acquisition or Import** </span>
Dataset can be in different formats (i.e. in `.csv`, `.xlsx` etc) and can be obtained from locally or remotely from an online repository. In this case, we will use `pandas` to read and import the data in `csv` format (comma seperated values).

First, let's make sure `pandas` is installed using the following command, then we will import the library using `import pandas as pd`.

<div class="alert alert-info alertinfo" style="margin-top: 20px">

**Note:**
- If you do not have `pandas` installed, just install it using by uncomment the code line below *(remove the hashtag #) and run `!pip istall pandas`

</div>

In [ ]:
# !pip install pandas

In [ ]:
import pandas as pd

### **<em style='color:darkred'> 1.1 Importing `csv` file using `read_csv()`</em>**

In this notebook example, we will use the automobile dataset from UCI Machine Learning repository (as in following URL). Let's try to click and access the following URL.

<a href="https://archive.ics.uci.edu/ml/machine-learning-databases/autos/imports-85.data" target='_blank'>`https://archive.ics.uci.edu/ml/machine-learning-databases/autos/imports-
85.data`</a>


To read the CSV file, we can use the `read_csv()` funciton from `pandas`. In the bracket, we pass in the file path of the file. The file path can be a physical file path or a remote URL address.

<div class="alert alert-info alertinfo" style="margin-top: 20px">
    
**Note:**
1. Notice that the data above, did not have any headers. Thus, in this case, in the `read_csv()` we also set the `headers` parameter to `None` so it will not use the first row as header automatically. We will add headers into it later!
2. Since the data to be read is from remote repository, always ensure internet is active!
</div>

In [ ]:
from google.colab import drive
drive.mount('/content/drive')

Drive already mounted at /content/drive; to attempt to forcibly remount, call drive.mount("/content/drive", force_remount=True).


In [ ]:
dataURL = 'https://archive.ics.uci.edu/ml/machine-learning-databases/autos/imports-85.data'

automobile = pd.read_csv(dataURL, header=None)

One the data is retrieved, we can check on the DataFrame using:
- `head(n)`function, where it will display the first `n` rows of data in the DataFrame.
- `tail(n)` function will show the last `n` rows.

By default, you can use both functions without the `n`, and the default setting is to display the first 5 or last 5 rows.

In [ ]:
automobile.head()

0    1            2    3    4     5            6    7      8     9   ...  \
0   3    ?  alfa-romero  gas  std   two  convertible  rwd  front  88.6  ...   
1   3    ?  alfa-romero  gas  std   two  convertible  rwd  front  88.6  ...   
2   1    ?  alfa-romero  gas  std   two    hatchback  rwd  front  94.5  ...   
3   2  164         audi  gas  std  four        sedan  fwd  front  99.8  ...   
4   2  164         audi  gas  std  four        sedan  4wd  front  99.4  ...   

    16    17    18    19    20   21    22  23  24     25  
0  130  mpfi  3.47  2.68   9.0  111  5000  21  27  13495  
1  130  mpfi  3.47  2.68   9.0  111  5000  21  27  16500  
2  152  mpfi  2.68  3.47   9.0  154  5000  19  26  16500  
3  109  mpfi  3.19  3.40  10.0  102  5500  24  30  13950  
4  136  mpfi  3.19  3.40   8.0  115  5500  18  22  17450  

[5 rows x 26 columns]

In [ ]:
automobile.tail()

0   1      2       3      4     5      6    7      8      9   ...   16  \
200  -1  95  volvo     gas    std  four  sedan  rwd  front  109.1  ...  141   
201  -1  95  volvo     gas  turbo  four  sedan  rwd  front  109.1  ...  141   
202  -1  95  volvo     gas    std  four  sedan  rwd  front  109.1  ...  173   
203  -1  95  volvo  diesel  turbo  four  sedan  rwd  front  109.1  ...  145   
204  -1  95  volvo     gas  turbo  four  sedan  rwd  front  109.1  ...  141   

       17    18    19    20   21    22  23  24     25  
200  mpfi  3.78  3.15   9.5  114  5400  23  28  16845  
201  mpfi  3.78  3.15   8.7  160  5300  19  25  19045  
202  mpfi  3.58  2.87   8.8  134  5500  18  23  21485  
203   idi  3.01  3.40  23.0  106  4800  26  27  22470  
204  mpfi  3.78  3.15   9.5  114  5400  19  25  22625  

[5 rows x 26 columns]

### **<em style='color:darkred'> 1.2 Adding headers to the data</em>**
<div class="alert alert-info alertinfo" style="margin-top: 20px">
    
**Note:** Based on the observation above, notice that without the headers, is difficult to understand the data!
</div>

Therefore, we will assign the names of the header manually based on information available in
<a href="https://archive.ics.uci.edu/dataset/10/automobile" target='_blank'>https://archive.ics.uci.edu/dataset/10/automobile</a>

We will replace the headers by assigning the information to the `columns` property of the DataFrame

In [ ]:
automobile.columns

Index([ 0,  1,  2,  3,  4,  5,  6,  7,  8,  9, 10, 11, 12, 13, 14, 15, 16, 17,
       18, 19, 20, 21, 22, 23, 24, 25],
      dtype='int64')

In [ ]:
headers = ["symboling","normalized-losses","make","fuel-type","aspiration", "num-of-doors","body-style",
         "drive-wheels","engine-location","wheel-base", "length","width","height","curb-weight","engine-type",
         "num-of-cylinders", "engine-size","fuel-system","bore","stroke","compression-ratio","horsepower",
         "peak-rpm","city-mpg","highway-mpg","price"]

automobile.columns = headers
automobile.head()

symboling normalized-losses         make fuel-type aspiration num-of-doors  \
0          3                 ?  alfa-romero       gas        std          two   
1          3                 ?  alfa-romero       gas        std          two   
2          1                 ?  alfa-romero       gas        std          two   
3          2               164         audi       gas        std         four   
4          2               164         audi       gas        std         four   

    body-style drive-wheels engine-location  wheel-base  ...  engine-size  \
0  convertible          rwd           front        88.6  ...          130   
1  convertible          rwd           front        88.6  ...          130   
2    hatchback          rwd           front        94.5  ...          152   
3        sedan          fwd           front        99.8  ...          109   
4        sedan          4wd           front        99.4  ...          136   

   fuel-system  bore  stroke compression-ratio horsepower  peak-rpm city-mpg  \
0         mpfi  3.47    2.68               9.0        111      5000       21   
1         mpfi  3.47    2.68               9.0        111      5000       21   
2         mpfi  2.68    3.47               9.0        154      5000       19   
3         mpfi  3.19    3.40              10.0        102      5500       24   
4         mpfi  3.19    3.40               8.0        115      5500       18   

  highway-mpg  price  
0          27  13495  
1          27  16500  
2          26  16500  
3          30  13950  
4          22  17450  

[5 rows x 26 columns]

We can always check the Data Type of the variable by using the `type()` function.

In [ ]:
type(automobile)

pandas.core.frame.DataFrame

### **<em style='color:darkred'> 1.3 Getting basic info of the DataFrame</em>**
In `pandas` DataFrame, there are some built-in functions and properties that we can use to get essential information.
For example:
- `dtypes`: shows the data type for each column
- `shape`: shows the size of the DataFrame in `(x, y)` format, `x` represents no. of rows, `y` shows the no. of columns.
- `columns`: shows the list of column names
- `index`: shows the row index
- `values`: show a list of row values.

In [ ]:
print(automobile.dtypes)

symboling              int64
normalized-losses     object
make                  object
fuel-type             object
aspiration            object
num-of-doors          object
body-style            object
drive-wheels          object
engine-location       object
wheel-base           float64
length               float64
width                float64
height               float64
curb-weight            int64
engine-type           object
num-of-cylinders      object
engine-size            int64
fuel-system           object
bore                  object
stroke                object
compression-ratio    float64
horsepower            object
peak-rpm              object
city-mpg               int64
highway-mpg            int64
price                 object
dtype: object


In [ ]:
print(automobile.shape)

(205, 26)


In [ ]:
print(automobile.columns)

Index(['symboling', 'normalized-losses', 'make', 'fuel-type', 'aspiration',
       'num-of-doors', 'body-style', 'drive-wheels', 'engine-location',
       'wheel-base', 'length', 'width', 'height', 'curb-weight', 'engine-type',
       'num-of-cylinders', 'engine-size', 'fuel-system', 'bore', 'stroke',
       'compression-ratio', 'horsepower', 'peak-rpm', 'city-mpg',
       'highway-mpg', 'price'],
      dtype='object')


In [ ]:
print(automobile.index)

RangeIndex(start=0, stop=205, step=1)


In [ ]:
print(automobile.values)

[[3 '?' 'alfa-romero' ... 21 27 '13495']
 [3 '?' 'alfa-romero' ... 21 27 '16500']
 [1 '?' 'alfa-romero' ... 19 26 '16500']
 ...
 [-1 '95' 'volvo' ... 18 23 '21485']
 [-1 '95' 'volvo' ... 26 27 '22470']
 [-1 '95' 'volvo' ... 19 25 '22625']]


### **<em style='color:darkred'> 1.4 Save the DataFrame to File</em>**
Especially for data retrieved from remote sources, once the dataset has been read, we can save the DataFrame into a file using `to_csv()` function and pass in the filenname as parameter.
<div class="alert alert-info alertinfo" style="margin-top: 20px">

**Note:**
- Meanwhile, set the parameter of `index` to `False` disable the index column so that the index of the data will not be save as a new column in the exported file.
</div>
<div class="alert alert-success alertsuccess" style="margin-top: 20px">

**Explore:**
- `pandas` also support export to multiple file types, including `to_excel()`, `to_json()`, `to_sql()`, `to_html()` etc.
</div>

Once export successful, you may find the exported file in the same folder with the notebook file.

In [ ]:
automobile.to_csv('automobile.csv', index=False)

<hr>

## <span style='color:darkred'> **STEP 2: Data Pre-Processing** </span>
In this step, the automobile data will be cleaned and transformed to the format that ready for analysis.

Let's see the data again.

In [47]:
automobile.head(10)

symboling  normalized-losses         make fuel-type aspiration  \
0          3              122.0  alfa-romero       gas        std   
1          3              122.0  alfa-romero       gas        std   
2          1              122.0  alfa-romero       gas        std   
3          2              164.0         audi       gas        std   
4          2              164.0         audi       gas        std   
5          2              122.0         audi       gas        std   
6          1              158.0         audi       gas        std   
7          1              122.0         audi       gas        std   
8          1              158.0         audi       gas      turbo   
9          2              192.0          bmw       gas        std   

  num-of-doors   body-style drive-wheels engine-location  wheel-base  ...  \
0          two  convertible          rwd           front        88.6  ...   
1          two  convertible          rwd           front        88.6  ...   
2          two    hatchback          rwd           front        94.5  ...   
3         four        sedan          fwd           front        99.8  ...   
4         four        sedan          4wd           front        99.4  ...   
5          two        sedan          fwd           front        99.8  ...   
6         four        sedan          fwd           front       105.8  ...   
7         four        wagon          fwd           front       105.8  ...   
8         four        sedan          fwd           front       105.8  ...   
9          two        sedan          rwd           front       101.2  ...   

   fuel-system  bore  stroke  compression-ratio horsepower peak-rpm    price  \
0         mpfi  3.47    2.68                9.0      111.0   5000.0  13495.0   
1         mpfi  3.47    2.68                9.0      111.0   5000.0  16500.0   
2         mpfi  2.68    3.47                9.0      154.0   5000.0  16500.0   
3         mpfi  3.19    3.40               10.0      102.0   5500.0  13950.0   
4         mpfi  3.19    3.40                8.0      115.0   5500.0  17450.0   
5         mpfi  3.19    3.40                8.5      110.0   5500.0  15250.0   
6         mpfi  3.19    3.40                8.5      110.0   5500.0  17710.0   
7         mpfi  3.19    3.40                8.5      110.0   5500.0  18920.0   
8         mpfi  3.13    3.40                8.3      140.0   5500.0  23875.0   
9         mpfi  3.50    2.80                8.8      101.0   5800.0  16430.0   

  horsepower-binned  city-L/100km  highway-L/100km  
0               Low     11.190476         8.703704  
1               Low     11.190476         8.703704  
2            Medium     12.368421         9.038462  
3               Low      9.791667         7.833333  
4               Low     13.055556        10.681818  
5               Low     12.368421         9.400000  
6               Low     12.368421         9.400000  
7               Low     12.368421         9.400000  
8            Medium     13.823529        11.750000  
9               Low     10.217391         8.103448  

[10 rows x 27 columns]

Now, we will go through some of the processing that we can do in Python, using the automobile dataset.
1. Checking and handling the missing values
2. Checking and handling the duplicates
3. Ensure the data in correct data types.
4. Generating new columns of data.
5. Transforming data into categorical (Binning)
6. Data Encoding (One Hot Encoding)

### **<em style='color:darkred'> 2.1 Missing Values</em>**
In real-world cases, data is not always perfect and often comes with missing values, that may be due to system or human error. These missing values must be handled properly to minimize the impact to the latter process.

<div class="alert alert-info alertinfo" style="margin-top: 20px">

**Note:**
- Based on the observation above, notice that, there are `?` exist in columns of `normalized-losses`. This could be a problem! because eventhough we knew `?` indicate that missing value, but for Python `?` is still a value.
</div>

Thus, to better represent the missing value, we will replace the `?` with `NaN` (not a number) from `numpy` library using the `replace()` function. In the `replace()` function, we will set the `inplace` value to `True` to make sure changes made to the DataFrame is permanent.

In [ ]:
import numpy as np

automobile.replace("?", np.nan, inplace=True)

In [ ]:
automobile.head(10)

symboling normalized-losses         make fuel-type aspiration num-of-doors  \
0          3               NaN  alfa-romero       gas        std          two   
1          3               NaN  alfa-romero       gas        std          two   
2          1               NaN  alfa-romero       gas        std          two   
3          2               164         audi       gas        std         four   
4          2               164         audi       gas        std         four   
5          2               NaN         audi       gas        std          two   
6          1               158         audi       gas        std         four   
7          1               NaN         audi       gas        std         four   
8          1               158         audi       gas      turbo         four   
9          0               NaN         audi       gas      turbo          two   

    body-style drive-wheels engine-location  wheel-base  ...  engine-size  \
0  convertible          rwd           front        88.6  ...          130   
1  convertible          rwd           front        88.6  ...          130   
2    hatchback          rwd           front        94.5  ...          152   
3        sedan          fwd           front        99.8  ...          109   
4        sedan          4wd           front        99.4  ...          136   
5        sedan          fwd           front        99.8  ...          136   
6        sedan          fwd           front       105.8  ...          136   
7        wagon          fwd           front       105.8  ...          136   
8        sedan          fwd           front       105.8  ...          131   
9    hatchback          4wd           front        99.5  ...          131   

   fuel-system  bore  stroke compression-ratio horsepower  peak-rpm city-mpg  \
0         mpfi  3.47    2.68               9.0        111      5000       21   
1         mpfi  3.47    2.68               9.0        111      5000       21   
2         mpfi  2.68    3.47               9.0        154      5000       19   
3         mpfi  3.19    3.40              10.0        102      5500       24   
4         mpfi  3.19    3.40               8.0        115      5500       18   
5         mpfi  3.19    3.40               8.5        110      5500       19   
6         mpfi  3.19    3.40               8.5        110      5500       19   
7         mpfi  3.19    3.40               8.5        110      5500       19   
8         mpfi  3.13    3.40               8.3        140      5500       17   
9         mpfi  3.13    3.40               7.0        160      5500       16   

  highway-mpg  price  
0          27  13495  
1          27  16500  
2          26  16500  
3          30  13950  
4          22  17450  
5          25  15250  
6          25  17710  
7          25  18920  
8          20  23875  
9          22    NaN  

[10 rows x 26 columns]

#### **<em style='color:darkred'> 2.1.1 Identify Missing Values</em>**

Now, in the case of the automobile dataset, the missing values are converted to Python's default. We can use following functions to identify these missing values:
- `isnull()`
- `isna()`

<div class="alert alert-info alertinfo" style="margin-top: 20px">

**Note:**
- in fact both `isnull()` and `isna()` are the same. You can check by `pd.isnull()==pd.isna()`
</div>

The output from both functions is a `bool` value indicating whether the value that is passed into the argument is in fact missing data.`True` stands for missing value, while `False` stands for not missing value.

In [ ]:
automobile.isna()

symboling  normalized-losses   make  fuel-type  aspiration  num-of-doors  \
0        False               True  False      False       False         False   
1        False               True  False      False       False         False   
2        False               True  False      False       False         False   
3        False              False  False      False       False         False   
4        False              False  False      False       False         False   
..         ...                ...    ...        ...         ...           ...   
200      False              False  False      False       False         False   
201      False              False  False      False       False         False   
202      False              False  False      False       False         False   
203      False              False  False      False       False         False   
204      False              False  False      False       False         False   

     body-style  drive-wheels  engine-location  wheel-base  ...  engine-size  \
0         False         False            False       False  ...        False   
1         False         False            False       False  ...        False   
2         False         False            False       False  ...        False   
3         False         False            False       False  ...        False   
4         False         False            False       False  ...        False   
..          ...           ...              ...         ...  ...          ...   
200       False         False            False       False  ...        False   
201       False         False            False       False  ...        False   
202       False         False            False       False  ...        False   
203       False         False            False       False  ...        False   
204       False         False            False       False  ...        False   

     fuel-system   bore  stroke  compression-ratio  horsepower  peak-rpm  \
0          False  False   False              False       False     False   
1          False  False   False              False       False     False   
2          False  False   False              False       False     False   
3          False  False   False              False       False     False   
4          False  False   False              False       False     False   
..           ...    ...     ...                ...         ...       ...   
200        False  False   False              False       False     False   
201        False  False   False              False       False     False   
202        False  False   False              False       False     False   
203        False  False   False              False       False     False   
204        False  False   False              False       False     False   

     city-mpg  highway-mpg  price  
0       False        False  False  
1       False        False  False  
2       False        False  False  
3       False        False  False  
4       False        False  False  
..        ...          ...    ...  
200     False        False  False  
201     False        False  False  
202     False        False  False  
203     False        False  False  
204     False        False  False  

[205 rows x 26 columns]

#### **<em style='color:darkred'> 2.1.2 Count Missing Values</em>**

To make it more easy to analyze, we can aggregate these information using `sum()` at the end of `isna()`. Since the missing values are represented as `True` (which is also 1) so by performing the `sum()`, any columns that return non-zero results are the columns that having missing values.

In [ ]:
print(automobile.isna().sum())

symboling             0
normalized-losses    41
make                  0
fuel-type             0
aspiration            0
num-of-doors          2
body-style            0
drive-wheels          0
engine-location       0
wheel-base            0
length                0
width                 0
height                0
curb-weight           0
engine-type           0
num-of-cylinders      0
engine-size           0
fuel-system           0
bore                  4
stroke                4
compression-ratio     0
horsepower            2
peak-rpm              2
city-mpg              0
highway-mpg           0
price                 4
dtype: int64


Based on the summary above, each column has 205 rows of data, seven columns containing missing data:

1. `normalized-losses`: 41 missing data
2. `num-of-doors`: 2 missing data
3. `bore`: 4 missing data
4. `stroke` : 4 missing data
5. `horsepower`: 2 missing data
6. `peak-rpm`: 2 missing data
7. `price`: 4 missing data

#### **<em style='color:darkred'> 2.1.3 Handling the Missing Values</em>**

<div class="alert alert-info alertinfo" style="margin-top: 20px">

<span style='color:black'>**How to deal with missing data?**</span>

1. drop data
   - drop the whole row
   - drop the whole column
2. replace data
   - replace it by mean
   - replace it by frequency
   - replace it based on other functions



</div>
Whole columns should be dropped only if most entries in the column are empty.
<br><br>
In our dataset, none of the columns are empty enough to drop entirely. We have some freedom in choosing which method to replace data; however, some methods may seem more reasonable than others.
<br>
<br>
We will handle the missing values differently on different columns:


<span style='color:brown'>**Replace by mean:**</span>
- `normalized-losses`: 41 missing data, replace them with mean
- `stroke`: 4 missing data, replace them with mean
- `bore`: 4 missing data, replace them with mean
- `horsepower`: 2 missing data, replace them with mean
- `peak-rpm`: 2 missing data, replace them with mean


Here, the beauty of programming language is the ability to make custom functions, same case here for Python.

Let's define a function that can be easily re-used for replacing missing values by mean.

<div class="alert alert-warning alertwarning" style="margin-top: 20px">

**Info:**
1. In the function below, we will convert the values in the targetted colum into `float` then call the `mean()` function. Setting `axis=0` will define the calculation to be done column-wise.
2. Then, will call the `replace()` function for that column to replace the NaN using the average values. setting `inplace=True` will make sure the changes are permanent.
</div>

In [ ]:
def replace_by_mean(data, col):
    avg = data[col].astype(float).mean(axis=0)
    data[col].replace(np.nan, avg, inplace=True)

Now, we will define the list of columns name, so that we can loop and process all the columns more easily.

In [ ]:
cols = ['normalized-losses','stroke','bore','horsepower','peak-rpm']
for c in cols:
    replace_by_mean(automobile, c)

Let's check again the missing values status in the dataset. You will notice most of the missing values are handled. There are two more columns remaining:
- `num-of-doors`
- `price`

In [ ]:
print(automobile.isna().sum())

symboling            0
normalized-losses    0
make                 0
fuel-type            0
aspiration           0
num-of-doors         2
body-style           0
drive-wheels         0
engine-location      0
wheel-base           0
length               0
width                0
height               0
curb-weight          0
engine-type          0
num-of-cylinders     0
engine-size          0
fuel-system          0
bore                 0
stroke               0
compression-ratio    0
horsepower           0
peak-rpm             0
city-mpg             0
highway-mpg          0
price                4
dtype: int64


There is no change on the total number of data, because we replaced the missing values with mean values for most of the columns.

In [ ]:
print(automobile.shape)

(205, 26)


<span style='color:brown'>**Replace by frequency:**</span>

- `num-of-doors`: 2 missing data, replace them with "four".
- **Reason:**
    - using the `value_counts()` function, we can see majority of the sedans is four doors. Since four doors is most frequent, it is most likely to occur


<div class="alert alert-info alertinfo" style="margin-top: 20px">

**Note:**
- `value_counts()` can be called from columns with categorical values, to show the distribution of the unique elements in the column.
</div>

In [ ]:
print(automobile['num-of-doors'].value_counts())

num-of-doors
four    114
two      89
Name: count, dtype: int64


In [ ]:
automobile['num-of-doors'].replace(np.nan, "four", inplace=True)

Let's check again the missing values status in the dataset. Now we left one more column remaining that having missing values, which is `price`

In [ ]:
print(automobile.isna().sum())

symboling            0
normalized-losses    0
make                 0
fuel-type            0
aspiration           0
num-of-doors         0
body-style           0
drive-wheels         0
engine-location      0
wheel-base           0
length               0
width                0
height               0
curb-weight          0
engine-type          0
num-of-cylinders     0
engine-size          0
fuel-system          0
bore                 0
stroke               0
compression-ratio    0
horsepower           0
peak-rpm             0
city-mpg             0
highway-mpg          0
price                4
dtype: int64


<span style='color:brown'>**Drop the whole row:**</span>

- `price`: 4 missing data, simply delete the whole row
- **Reason:**
    - `price` is what we want to predict. Any data entry without price data cannot be used for prediction; therefore any row now without price data is not useful to us

To remove the row, we can use `dropna()` function, then in the `subset` parameter, pass in the name of the desired columns, `axis` set to `0` indicates operation done by row (column wise), and `inplace` to `True` to ensure changes are permanent.


In [ ]:
automobile.dropna(subset=['price'], axis=0, inplace=True)

After dropping the rows, usually the index wont be updated automatically, we need to invoke `reset_index()` after the `drop_na()`, setting `drop` parameter to `True` to avoid the function saving the previous index as a new columns in the DataFrame

In [ ]:
automobile.reset_index(drop=True, inplace=True)

Let's check again the missing values status in the dataset.

In [ ]:
print(automobile.isna().sum())

symboling            0
normalized-losses    0
make                 0
fuel-type            0
aspiration           0
num-of-doors         0
body-style           0
drive-wheels         0
engine-location      0
wheel-base           0
length               0
width                0
height               0
curb-weight          0
engine-type          0
num-of-cylinders     0
engine-size          0
fuel-system          0
bore                 0
stroke               0
compression-ratio    0
horsepower           0
peak-rpm             0
city-mpg             0
highway-mpg          0
price                0
dtype: int64


In [ ]:
automobile.shape

(201, 26)

### **<em style='color:darkred'> 2.2 Correction of Data Types</em>**
The last step in cleaning is to check and make sure all data is in the correct type.

In `pandas` we use:

- `.dtypes` to check the data type
- `astype()` to change the data type, where the parameter is the data type to change to.

In this case of automobile dataset, let's check again the data type

<div class="alert alert-warning alertwarning" style="margin-top: 20px">

**Note:**
- Python supports dynamic typing, where the data type is assigned based on the values.
- In general, numerical data will be either `int` or `float`, text will be `object`.
</div>

Notice that our dataset, some of the numerical columns are in the type of `object`, this is because initially these columns having missing values, which means not all values are numerical, thus, Python will automatically assign the column as `object`

In [ ]:
print(automobile.dtypes)

symboling              int64
normalized-losses     object
make                  object
fuel-type             object
aspiration            object
num-of-doors          object
body-style            object
drive-wheels          object
engine-location       object
wheel-base           float64
length               float64
width                float64
height               float64
curb-weight            int64
engine-type           object
num-of-cylinders      object
engine-size            int64
fuel-system           object
bore                  object
stroke                object
compression-ratio    float64
horsepower            object
peak-rpm              object
city-mpg               int64
highway-mpg            int64
price                 object
dtype: object


Since we have handled the missing values, let's change some of the numerical columns back to numerical, these columns include:
- `normalized-losses`
- `horsepower`
- `peak-rpm`
- `bore`
- `stroke`
- `price`

Let's convert these columns into float. To change the data type of the column, we can use `astype()` function and pass in the data type to change to, then assign back to the same column. This is to make the changes permanent by overwrite the original column.

<div class="alert alert-info alertinfo" style="margin-top: 20px">

**Note:**
- Not all columns can be converted to float, columns with string data will generate error when trying to convert to `float` type.
</div>


In [ ]:
cols = ['normalized-losses', 'horsepower', 'peak-rpm', 'bore', 'stroke', 'price']

automobile[cols] = automobile[cols].astype(float)

In [ ]:
automobile.dtypes

symboling              int64
normalized-losses    float64
make                  object
fuel-type             object
aspiration            object
num-of-doors          object
body-style            object
drive-wheels          object
engine-location       object
wheel-base           float64
length               float64
width                float64
height               float64
curb-weight            int64
engine-type           object
num-of-cylinders      object
engine-size            int64
fuel-system           object
bore                 float64
stroke               float64
compression-ratio    float64
horsepower           float64
peak-rpm             float64
city-mpg               int64
highway-mpg            int64
price                float64
dtype: object

### **<em style='color:darkred'> 2.3 Data Binning</em>**
**Why binning?**

Binning is a process of transforming continuous numerical variables into discrete categorical 'bins', for grouped analysis.

**Example:**

In the automobile dataset, `horsepower` is a real valued variable ranging from 48 to 288, it has 57 unique values. What if we only care about the price difference between cars with high horsepower, medium horsepower, and little horsepower (3 types)? Can we rearrange them into three ‘bins' to simplify analysis?

We will use the Pandas method `cut()` to segment the `horsepower` column into 3 bins

In [ ]:
automobile.horsepower

0      111.0
1      111.0
2      154.0
3      102.0
4      115.0
       ...  
196    114.0
197    160.0
198    134.0
199    106.0
200    114.0
Name: horsepower, Length: 201, dtype: float64

The following custom function will perform the binning

1. We would like 3 bins of equal size bandwidth so we use `numpy`'s `linspace(start_value, end_value, numbers_generated)` function.

2. Since we want to include the minimum value of horsepower we want to set `start_value=min(automobile["horsepower"])`.

3. Since we want to include the maximum value of horsepower we want to set `end_value=max(automobile["horsepower"])`.

4. Since we are building 3 bins of equal length, there should be 4 dividers, so `numbers_generated=4`.

5. Then, we set the group names for each of the 3 groups, and they apply the `cut()` function to divide the data of `horsepower` into the corresponding category.

In [ ]:
def get_bins(col_data, grp_name, num_bin):
    if col_data.dtypes != "object":
        bins = np.linspace(min(col_data), max(col_data),  num_bin)
        res  = pd.cut(col_data, bins, labels=grp_name, include_lowest=True)
        return res
    else:
        print('sorry, the column is not numerical')

In [ ]:
group_names = ['Low', 'Medium', 'High']

values = get_bins(automobile['horsepower'], group_names , 4)

automobile['horsepower-binned'] = values

In [ ]:
automobile.head()

symboling  normalized-losses         make fuel-type aspiration  \
0          3              122.0  alfa-romero       gas        std   
1          3              122.0  alfa-romero       gas        std   
2          1              122.0  alfa-romero       gas        std   
3          2              164.0         audi       gas        std   
4          2              164.0         audi       gas        std   

  num-of-doors   body-style drive-wheels engine-location  wheel-base  ...  \
0          two  convertible          rwd           front        88.6  ...   
1          two  convertible          rwd           front        88.6  ...   
2          two    hatchback          rwd           front        94.5  ...   
3         four        sedan          fwd           front        99.8  ...   
4         four        sedan          4wd           front        99.4  ...   

   fuel-system  bore  stroke  compression-ratio horsepower peak-rpm  city-mpg  \
0         mpfi  3.47    2.68                9.0      111.0   5000.0        21   
1         mpfi  3.47    2.68                9.0      111.0   5000.0        21   
2         mpfi  2.68    3.47                9.0      154.0   5000.0        19   
3         mpfi  3.19    3.40               10.0      102.0   5500.0        24   
4         mpfi  3.19    3.40                8.0      115.0   5500.0        18   

  highway-mpg    price  horsepower-binned  
0          27  13495.0                Low  
1          27  16500.0                Low  
2          26  16500.0             Medium  
3          30  13950.0                Low  
4          22  17450.0                Low  

[5 rows x 27 columns]

### **<em style='color:darkred'> 2.4 Creating New Columns</em>**
Data is usually collected from different agencies with different formats.

*Transform mpg to L/100km:*

In our dataset, the fuel consumption columns `city-mpg` and `highway-mpg` are represented by mpg (miles per gallon) unit. Assume we are developing an application in a country that accept the fuel consumption with L/100km standard. We will need to transform mpg into L/100km

The formula for unit conversion is
   
    L/100km = 235 / mpg

We can do many mathematical operations directly in `pandas`.

<div class="alert alert-info alertinfo" style="margin-top: 20px">

**Note:**
- In `pandas` we can add new column simply by assigning values into non-existing column.
- While assign values to existing column will replace them.
- If want to add new column at specific position, can use `insert()` function of the DataFrame. Example: `df.insert(2, "new_col" , data)` where `2` refer the position, `new_col` is the new column name, and `data` is the values to be assigned to the column.
</div>

In [ ]:
city    = 235/automobile['city-mpg']
highway = 235/automobile['highway-mpg']

In [ ]:
automobile['city-L/100km'] = city
automobile['highway-L/100km'] = highway

Then, we will remove the initial columns of `city-mpg` and `highway-mpg`

In [ ]:
automobile.drop(['city-mpg','highway-mpg'], axis=1, inplace=True)

In [ ]:
automobile.columns

Index(['symboling', 'normalized-losses', 'make', 'fuel-type', 'aspiration',
       'num-of-doors', 'body-style', 'drive-wheels', 'engine-location',
       'wheel-base', 'length', 'width', 'height', 'curb-weight', 'engine-type',
       'num-of-cylinders', 'engine-size', 'fuel-system', 'bore', 'stroke',
       'compression-ratio', 'horsepower', 'peak-rpm', 'price',
       'horsepower-binned', 'city-L/100km', 'highway-L/100km'],
      dtype='object')

### **<em style='color:darkred'> 2.5 Export Clean Data</em>**
Now, since the automobile data has been processed extensively, let's save the clean data into another file named `automobile_cleaned.csv` and set the `index` parameter to `False`

In [ ]:
automobile.to_csv('automobile_cleaned.csv', index=False)

<hr>